### DISCUSSION TITLE
Can I create my own button which will handle both the combinator selection and Rule addition functionality ?

### DISCUSSION THREAD

**Discussion by keerthankrishna** [react-querybuilder/react-querybuilder#420] [Q&A] [answered]
Is it possible for me to create buttons of my own which will function just like "+Rule" but instead of adding a Rule and a Combinator seperately, would behave like an "And Rule" button and a "Or Rule" button which when clicked  will directly add an "And selection line" or an "Or selection line respectively ?
![image](https://user-images.githubusercontent.com/24373382/207515898-457a6135-a463-44da-94ea-817b22717943.png)

**Comment by jakeboone02** [comment, marked as answer]
Hi, @keerthankrishna!

If you want to keep the combinators at the group level (which is the default), your "+ AND rule"/"+ OR rule" buttons would actually need to be "+ AND group"/"+ OR group". This CodeSandbox should point you in the right direction: https://codesandbox.io/s/infallible-albattani-q8wrx4?file=/src/App.tsx.

But it actually sounds like you want to use `independentCombinators`, which puts a combinator selector between each rule. That way you can have a query like this:

![image](https://user-images.githubusercontent.com/366438/207686290-dd42b19b-fda3-46c4-8bb9-002153f563a1.png)

and add an "AND" rule like this:

![image](https://user-images.githubusercontent.com/366438/207686400-2bd7e769-017e-460f-9820-dce4e388ce9a.png)

or add an "OR" rule like this:

![image](https://user-images.githubusercontent.com/366438/207686531-1c93350c-bdc1-4266-9586-ab321c150732.png)

If that's what you're looking for, there are ways to accomplish what you want with RQB as of v5.2.0 but they're a little convoluted. Let me know if you need an immediate solution.

If you can wait a bit (maybe a few days? I'm not sure when the next version will be released), v5.2.1 will support a relatively simple solution. You can try it out by clicking the link below, which opens a CodeSandbox that uses the latest RQB build from PR #421 and demonstrates the desired functionality.

https://codesandbox.io/s/react-querybuilder-test-for-combinatorpreceding-9149po?file=/src/App.tsx

Relevant code:

```tsx
// This replaces the default "+Rule" button with two separate buttons
const AddRules = (props: ActionWithRulesAndAddersProps) => (
  <>
    <button onClick={e => props.handleOnClick(e, 'and')}>+ AND Rule</button>
    <button onClick={e => props.handleOnClick(e, 'or')}>+ OR Rule</button>
  </>
);

// These functions take the "context" parameter and assign
// it to the (new) "combinatorPreceding" property
const onAddRule = (
  r: RuleType,
  _p: number[],
  _q: any,
  combinatorPreceding: string
) => ({
  ...r,
  combinatorPreceding,
});
const onAddGroup = (
  rg: RuleGroupTypeIC,
  _p: number[],
  _q: any,
  combinatorPreceding: string
) => ({
  ...rg,
  combinatorPreceding,
});
```

Those then get assigned to the `<QueryBuilder />` like this:

```tsx
<QueryBuilder
  independentCombinators
  query={queryIC}
  onQueryChange={q => setQueryIC(q)}
  controlElements={{ addRuleAction: AddRules }}
  onAddRule={onAddRule}
  onAddGroup={onAddGroup}
/>
```

**Reply by keerthankrishna** [reply]
@jakeboone02  Thank you so much for taking the time and answering my question in detail. I'm not in a hurry, so I can wait - Thank you so much for sharing the codesandbox link with one of the solutions. That really helps giving me few more options which I will explore further. Will reach out in this discussion chat if I have more questions on this. And yes, I'm looking to use this with the _independentCombinators_ but with your solution, I could even just keep it false (the _independentCombinators_) and have the custom button handle the AND/OR rule addition.

**Two more questions:**
_(1)_ is it possible for me to customize the AND/OR dropdown similar to how the Rule/Group buttons can be customized ? See below screenshot (it's an edited one but I'm looking to customize the library this way but I was not able to find customization options for the AND/OR dropdown. I would like to use the same independentCombinators config but instead of it being in a seperate line, I would like it to be in the same rule selection line.
![image](https://user-images.githubusercontent.com/24373382/207719328-de4ed9bd-8676-4220-9cb0-aa4199a141f6.png)


_(2)_ is it possible to disable GROUP button after one addition of a group ? Example: Once I click on GROUP, i should not be able to add any more groups anywhere. So both the Parent Group should be disabled and the Child Group should be disabled. I think you allow customization on Child groups being disabled via class but I think on the parent side, I don't see anything mentioned especially with disabling after addition of one group

**Reply by jakeboone02** [reply]
I'm not clear on what you're saying in that first paragraph w/r/t `independentCombinators` being `false` but still adding AND/OR to the query at the rule level.

As for your other questions:

1. You can achieve something close to that with CSS. You won't be able to insert an inline combinator control within the rule controls like you have pictured (the combinator control is between the value editor and the remove rule button) without using a custom `ruleGroup` component, but you can reorder the components within a rule.  The following CSS puts inline combinators to the right of the rule above them:
    ```css
    .ruleGroup-body {
      display: grid;
      grid-template-columns: 1fr auto;
      gap: 0.5rem;
      align-items: center;
    }
    ```
    ![image](https://user-images.githubusercontent.com/366438/207725151-2363f9c7-b4ed-4268-b721-a05131ea6d91.png)
2. You can enforce a "one group only" constraint using the `validator` prop and some CSS.
    Component:
    ```tsx
    // If any of the sub-rules are groups, return false (i.e. "invalid")
    const validator: QueryValidator = q =>
      !q.rules.some(r => typeof r !== 'string' && 'rules' in r);

    export const App = () => (
      <QueryBuilder validator={validator} />
    );
    ```
    CSS:
    ```css
    /* Hide any "+Group" buttons when the query is invalid */
    .queryBuilder-invalid .ruleGroup-addGroup {
      display: none;
    }
    ```

**Reply by keerthankrishna** [reply]
Thank you so much for this solution! I did not know I could style it this way to answer my (1)st question. Thanks for the options you gave me for my (2)nd question. Please disregard my first statement in my previous answer - I just wanted to say, your solution will help me handle ways to resolve my problem.
